Bundling 2.2.4

kandy opened the pull request.

Commit:
rename underscore methods

Commit:
rename underscore methods

adeelq committed:
#53370 EDC Coupon Codes - expiry days value of 0 sets coupon expiration date/time to coupon creation date/time

adeelq committed:
#53370 EDC Coupon Codes - expiry days value of 0 sets coupon expiration date/time to coupon creation date/time

adeelq committed:
Customer does not get reset when making order #56640

adeelq committed:
Customer does not get reset when making order #56640

Commit:
update version

Commit:
update version

Commit:
Merge branch 'testing' of https://github.com/dotmailer/dotmailer-magento2-extension into testing

adeelq committed:
insert already set $where variable into query

adeelq committed:
build single query

Commit:
force the default type for datafields value to match the type

Commit:
force the default type for datafields value to match the type

adeelq committed:
make wishlist_imported column clear with option Imported / Not Imported

adeelq committed:
make wishlist_imported column clear with option Imported / Not Imported

Commit:
open the link in a new tab

adeelq committed:
Clear config cache on last step on trial process

simon-letch committed:
Merge branch 'WishlistReport' of https://dotmailer.visualstudio.com/dm/_git/dotmailer%20for%20Magento%202 into WishlistReport

simon-letch committed:
Merge branch 'datafieldsDefaultValue' of https://dotmailer.visualstudio.com/dm/_git/dotmailer%20for%20Magento%202 into datafieldsDefaultValue

adeelq committed:
Merged PR 4429: insert already set $where variable into query

insert already set $where variable into query

adeelq committed:
Merge branch 'testing' of https://dotmailer.visualstudio.com/dm/_git/dotmailer%20for%20Magento%202 into TrialProcess

adeelq committed:
Clear config cache on automap #57148

adeelq committed:
Merged PR 4448: Merge WishlistReport to testing

#57156

Commit:
Merged PR 4463: open the link in a new tab

open the link in a new tab

Related work items: #57172

Commit:
#57150 move to helper and replace the datetime check

Commit:
Merge branch 'datafieldsDefaultValue' of ssh://vs-ssh.visualstudio.com:22/dm/_ssh/dotmailer%20for%20Magento%202 into datafieldsDefaultValue

adeelq committed:
Merged PR 4460: Clear config cache on automap #57148

Clear config cache on automap #57148

Related work items: #57148

adeelq committed:
Merged PR 4459: Merge TrialProcess to testing

#57148

Commit:
Merged PR 4446: Merge datafieldsDefaultValue to testing

Related work items: #57159

adeelq committed:
create plugin for interface instead of concrete class

adeelq committed:
Change "Ddg_CustomerManagementPlugin" to create plugin on interface #57343

adeelq committed:
Change "ddg_customer_email_disabler" to create plugin on interface #57347

adeelq committed:
bump it up

adeelq committed:
bump it up

adeelq committed:
bump it up

adeelq committed:
Merged PR 4516: Change "Ddg_CustomerManagementPlugin" to create plugin on interface #57343

Change "Ddg_CustomerManagementPlugin" to create plugin on interface #57343

Related work items: #57343

adeelq committed:
Merged PR 4515: Change "ddg_customer_email_disabler" to create plugin on interface #57347

Change "ddg_customer_email_disabler" to create plugin on interface #57347

Related work items: #57347

adeelq committed:
Merge branch 'testing' of https://dotmailer.visualstudio.com/dm/_git/dotmailer%20for%20Magento%202 into customerEmailDisabler

# Conflicts:
#	composer.json
#	etc/module.xml

adeelq committed:
Merge branch 'testing' of https://dotmailer.visualstudio.com/dm/_git/dotmailer%20for%20Magento%202 into TransactionalEmail

# Conflicts:
#	composer.json
#	etc/module.xml

adeelq committed:
Merged PR 4511: create plugin for interface instead of concrete class

create plugin for interface instead of concrete class #57274 #57273

Related work items: #57274

adeelq committed:
extra checks around catalog sync and registering item in importer

adeelq committed:
Add logging to importer and catalog sync

adeelq committed:
change from bulk to single for single type

adeelq committed:
Merged PR 4533: extra checks around catalog sync and registering item in importer #57156

extra checks around catalog sync and registering item in importer #57156

simon-letch committed:
Merged PR 4526: Only delete dir if $path is set and it contains ‘var’

Only delete dir if $path is set and it contains ‘var’

Ported from M1

Related work items: #56698

Commit:
version number upgrade script for abandoned cart

Commit:
version number upgrade script for abandoned cart

simon-letch committed:
Merged PR 4544: Merge copy-fixes to testing

BUNDLE-343
BUNDLE-335
BUNDLE-334
BUNDLE-333

Copy fixes

adeelq committed:
Catalog report mass delete action throws exception #57425 #57520

adeelq committed:
Merge branch 'testing' of https://dotmailer.visualstudio.com/dm/_git/dotmailer%20for%20Magento%202 into CatalogReport

# Conflicts:
#	composer.json
#	etc/module.xml

Commit:
Merged PR 4558: version number upgrade script for abandoned cart

version number upgrade script for abandoned cart

adeelq committed:
Merged PR 4572: Merge CatalogReport to testing

Related work items: #57425, #57520

adeelq committed:
bump version number

adeelq committed:
bump version number

adeelq committed:
Merged PR 4597: bump version number #57575

bump version number #57575

adeelq committed:
update release notes

adeelq committed:
Merged PR 4598: Merge Release237 to master

#57575

simon-letch committed:
Merged PR 4594: Remove mailcheck feature

Remove mailcheck feature

adeelq committed:
fix for encoding issue #57968

adeelq committed:
fix for encoding issue #57968

simon-letch committed:
Merged PR 4645: BUNDLE-333

BUNDLE-333

Correct case

adeelq committed:
change strlen to mbstrlen #57968

adeelq committed:
Merge branch 'testing' of https://dotmailer.visualstudio.com/dm/_git/dotmailer%20for%20Magento%202 into Encoding

adeelq committed:
Merged PR 4647: fix for encoding issue #57968

fix for encoding issue #57968

Related work items: #57968

simon-letch committed:
Merged PR 4641: Merge sl/bundle-330 to testing

Add event listener to enable cross frame communication.

simon-letch committed:
Merged PR 4657: Merge sl/copy-fixes to testing

adeelq committed:
Dotmailer login screen is shown in Automation Studio #57384

adeelq committed:
Merge branch 'testing' of https://dotmailer.visualstudio.com/dm/_git/dotmailer%20for%20Magento%202 into oAuth

adeelq committed:
Remove curl options #57384

simon-letch committed:
getAccessToken returns accessToken. We were overwriting refreshToken on the admin user with the access token. Response token is created and set only on initial oauth2 exchange of authorisation code. Changed to return the accessToken.

adeelq committed:
Merged PR 4663: Merge oAuth to testing

Related work items: #57384

adeelq committed:
Merged PR 4681: Translation

Translation

adeelq committed:
Merged PR 4548: Merge OrderReview to testing

Related work items: #57341

adeelq committed:
register with importer if either there is either importData or a file #58204

simon-letch committed:
Merged PR 4694: Check if directory to be deleted contains the var directory defined in Magento.

Check if directory to be deleted contains the var directory defined in Magento.

This resolves potential problems with non linux environments and makes the check far stricter.

adeelq committed:
Merged PR 4697: register with importer if either there is either importData or a file #58204

register with importer if either there is either importData or a file #58204

Related work items: #58204

adeelq committed:
Merged PR 4687: replace substr with mb_substr for multi byte safe string split #58028

replace substr with mb_substr for multi byte safe string split #58028

Related work items: #58028

simon-letch committed:
Better logging for failed IP address validation

adeelq committed:
Merged PR 4804: observer delete_before #58709

BUNDLE-436

Related work items: #58709

Commit:
Merged PR 4805: Bundle - 466 subscriber not added to subscriber address book when resubscribes

Bundle - 466 subscriber not added to subscriber address book when resubscribes

Related work items: #58716

Commit:
BUNDLE-351 wishlists not syncing

simon-letch committed:
Merged PR 4755: Better logging for failed IP address validation

Better logging for failed IP address validation

Commit:
Merged PR 4814: BUNDLE-351 wishlists not syncing

BUNDLE-351 wishlists not syncing

Related work items: #58739

adeelq committed:
Merged PR 4794: Merge AQ/FirstOrderAutomation to testing

So this will prevent duplicates at any given time. Since customer should only be enrolled once to programme 'first_order_automation'.

Related work items: #58626

simon-letch committed:
Merged PR 4838: Change plural to singular

Change plural to singular

BUNDLE-470

Commit:
send transactional emails

simon-letch committed:
Merged PR 4825: Only run job if we haven't already ran it

Schedule->tryLockJob ensures that we have only one running job of any type at any one time. This applies to duplicate messages in teh cron_schedule table, multiple users running cron etc.

We have instances where we get multiple schedules with the same scheduled_at date. We only want to process one of them.

This PR gets the currently executing job, then checks if a job of the same type and scheduled_at date has already been ran. If so, it exits, else it processes the job.

I think i've wrote it according to Magento standards. I'm more than happy to be proved wrong.

Commit:
Merged PR 4830: BUNDLE-350 Magento guests' contacts aren't imported into the 'Guests' address

BUNDLE-350 Magento guests' contacts aren't imported into the 'Guests' address book in Dotmailer

Commit:
remove unsued code

Commit:
create configuration for email templates

adeelq committed:
Merged PR 4844: Only make api call to deleteAddressBookContact if is a contact id #58865

Only make api call to deleteAddressBookContact if is a contact id #58865

Related work items: #58865

adeelq committed:
API Credentials - Value in config for connector/api/endpoint is saved for website scope and not default scope #53299

adeelq committed:
revert origin function. create seperate function to get config for selected scope in admin #53299

adeelq committed:
Merged PR 4855: API Credentials - Value in config for connector/api/endpoint is saved for

API Credentials - Value in config for connector/api/endpoint is saved for website scope and not default scope #53299

Related work items: #53299

simon-letch committed:
Merged PR 4887: Use getStatuses to prevent error on upgrade 2.2.2

Use getStatuses to prevent error on upgrade 2.2.2

adeelq committed:
Merged PR 4890: Check for returned error. mark item failed and save error msg to item #59110

Check for returned error. mark item failed and save error msg to item #59110

Related work items: #59110

adeelq committed:
Merged PR 4897: Custom OAUTH Domain instead of Custom OAUTH domain.

Custom OAUTH Domain instead of Custom OAUTH domain.

adeelq committed:
Merged PR 4883: register with importer after contact is saved and have an id #58953

register with importer after contact is saved and have an id #58953

Related work items: #58953

Commit:
Merge branch 'testing' into cleaning

# Conflicts:
#	Model/Apiconnector/Client.php

Commit:
fix merging conflicts

Commit:
Merged PR 4845: remove unsued code

remove unsued code

adeelq committed:
Add config #59168

adeelq committed:
Add default value for config #59187

adeelq committed:
Sync subscriber if sales data if enable in config #59169

adeelq committed:
Merge branch 'testing' of https://dotmailer.visualstudio.com/dm/_git/dotmailer%20for%20Magento%202 into AQ/saveEndpointMethodOne

adeelq committed:
use admin website on validation #58887

adeelq committed:
BUNDLE-488 Wishlists count is decreased in dotmailer upon adding/deleting products from wishlists in magento	#59262

adeelq committed:
Merged PR 4919: Merge AQ/SubscriberSalesData to testing

Related work items: #59168, #59169, #59187

adeelq committed:
Merged PR 4940: BUNDLE-488 Wishlists count is decreased in dotmailer upon adding/deleting products from wishlists in magento #59262

So this one is related to https://github.com/dotmailer/dotmailer-magento2-extension/commit/f4e41f63041c94dd7a8fbb136bef74ac8ccadef4.

Calin pointed
->addStoreFilter($this->getSharedStoreIds()) you don’t have this ids in a cron mode
wishlist items will be empty all the time

Found another bug in RemoveWishlist.php which will not remove wishlist from account. Minor fix.

Related work items: #59262

adeelq committed:
Force save api endpoint on api validation. only save it once #58887

adeelq committed:
Merge branch 'testing' of https://dotmailer.visualstudio.com/dm/_git/dotmailer%20for%20Magento%202 into AQ/saveEndpointMethodOne

adeelq committed:
Merged PR 4929: Merge AQ/saveEndpointMethodOne to testing

Related work items: #53299, #58887

Commit:
Merged PR 4924: guests email been duplicated in email contact

guests email been duplicated in email contact

Commit:
Merged PR 4951: Merge prefixForTables to testing

Commit:
proccess the abandoned carts when the first one is disable

adeelq committed:
Subscriber status not updated in Magento when from DM account #59249

adeelq committed:
Subscriber status not updated in Magento when from DM account #59249

Commit:
config saving/reseting the templates, new api request for campaign

Commit:
config saving/reseting the templates, new api request for campaign

(cherry picked from commit 79c2611)

Commit:
email templates - run button in admin for sync

Commit:
email templates - run button in admin for sync

(cherry picked from commit d367954)

Commit:
update test case

simon-letch committed:
Only add if not in array to prevent duplicates appearing in multi site scenarios.

Commit:
Revert "email templates - run button in admin for sync"

This reverts commit f555d41

Commit:
Revert "config saving/reseting the templates, new api request for campaign"

This reverts commit 7d9700a

Commit:
revert changes

Commit:
fix conflict after revert

Commit:
Merged PR 4977: proccess the abandoned carts when the first one is disable

proccess the abandoned carts when the first one is disable

simon-letch committed:
Add braces to comply with styling rules

adeelq committed:
Merged PR 4981: Subscriber status not updated in Magento when from DM account #59249

Subscriber status not updated in Magento when from DM account #59249

Related work items: #59249

Commit:
transactional emails - config for templates

Commit:
transactional emails - cron to sync the templates every 6h

Commit:
Merged PR 5021: transactional emails - cron to sync the templates every 6h

transactional emails - cron to sync the templates every 6h

Commit:
Merged PR 5022: transactional emails - config for templates

transactional emails - config for templates

Commit:
new section for email templates

Commit:
rename template observer

Commit:
transactional emails - newsletter success

simon-letch committed:
Merged PR 5054: Check protocol current page was requested over to determine which protocol

Check protocol current page was requested over to determine which protocol emailcapture requests should be made over.

This means if a store is configured to use insecure front end urls, but in fact serves secure requests, we correctly capture the email.

Rather than getting insecure request errors due to the page being https and the request to emailcapture being http.

#61840

Related work items: #61840

adeelq committed:
Merged PR 5047: Can't save a condition in exclusion rules #58341

Can't save a condition in exclusion rules #58341

Related work items: #58341

Commit:
transactional emails - fix the config values that inherits default values

Commit:
transactional emails - new order confirmation

Commit:
transactional emails - clean config cache after saving new values

Commit:
transactional emails - order confirmation for guests

Commit:
transactional emails - new shipment

Commit:
transactional emails - new shipment for guest

Commit:
transactional emails - new account confirmed

Commit:
transactional emails - new subscription confirmation

Commit:
transactional emails - unsubscribe success

Commit:
transactional emails - order update

Commit:
transactional emails - order update for guest

Commit:
transactional email - forgot password for user and customer

Commit:
transactional emails - invoice update

Commit:
transactional emials - invoice update guest

Commit:
transactional emails - new invoice

Commit:
transactional emails - new invoice guest

Commit:
transactional emails - new credit memo

Commit:
transactional emails - new credit guest

Commit:
transactional emails - contact form

Commit:
transactional emails - credit memo update

Commit:
transactional emails - credit memo update guest

Commit:
transactional emails - remind password

Commit:
transactional email - send product to friend

Commit:
transactional emails - wishlist sharing

Commit:
transactional emails - shipment update for guest

Commit:
remove the logging

Commit:
fix typo

Commit:
change order for sales template in config

Commit:
Merged PR 5149: fix abandoned cart to send campaign if is enabled and mapped

fix abandoned cart to send campaign if is enabled and mapped

Related work items: #58746

adeelq committed:
Merged PR 5159: Merge AQ/V3.0.0_Release to testing

updated version number and readme

Related work items: #62056

adeelq committed:
Merged PR 5164: Merge AQ/v2.4.0Release to testing

Related work items: #62056

adeelq committed:
Merge branch 'master' into testing

# Conflicts:
#	README.md
#	composer.json
#	etc/module.xml

simon-letch committed:
Merged PR 5167: Resolve merge conflicts

Merge master -> testing to resolve merge conflicts

Related work items: #57575

adeelq committed:
Merge branch 'testing' of https://dotmailer.visualstudio.com/dm/_git/dotmailer%20for%20Magento%202 into testing

adeelq committed:
Merged PR 5171: Merge AQ/TmpTesting to testing

Related work items: #57575

adeelq committed:
Merge branch 'testing' of https://dotmailer.visualstudio.com/dm/_git/dotmailer%20for%20Magento%202 into testing

simon-letch committed:
Correctly mark campaigns sent via dotmailer as in 'Processing' state

Previously we weren't sending in batches of 10. We were also not marking sent campaigns as 'processing' meaning we processed them multiple times. To repro:

Insert more than 10 records into db. use your gmail address and increment the alias part. Set an appropriate campaign id too.

insert into magento_debug_20.email_campaign (campaign_id, email, customer_id, sent_at, order_increment_id, quote_id, message, checkout_method, store_id, event_name, send_id, send_status, created_at, updated_at)
values (campaign_id_of_your_choice, 'youremailaddres+1@gmail.com', 9, null, '', 0, '', '', 1, 'test', '', 0, '2017-12-08 17:45:24', null);

Wait for cron to run.

We expect to see top 10 marked as processing, but we do not.

After the fix, we see 10 are marked as processing. Also, the checksendstatus also marks all successfully sent sends as sent now.

I've removed all usages of implode when passing it into an in query. I need some help testing the installdata and contact code...

Related work items: #62374

simon-letch committed:
v2.4.1 Release

Related work items: #62374

Commit:
fix space phpcs

Commit:
Merge branch 'testing' into transactionalEmails

# Conflicts:
#	Model/Cron.php
#	Model/ResourceModel/Cron/Collection.php

simon-letch committed:
Merge branch 'master' into testing

simon-letch committed:
Merge branch 'master' into testing

adeelq committed:
Merged PR 5185: Security - encrypt the passwords when saving to database #57864

Security - encrypt the passwords when saving to database #57864

# Conflicts:
#	etc/module.xml

Related work items: #57864

simon-letch committed:
Merged PR 5309: Return value according to whether the resubscribe was successful

Resubscribe messages get 'stuck' in the importer table and are getting processed every importer cron run. The status of the record remains not_imported and never gets set to success/failure.

####The fix
Return true/false. The return of this method gets checked by isset an if so, the importer record is marked as success/failed depending on the result.

If isset is false, then the importer record remains as NOT_IMPORTED and gets picked up again the next time the importer cron runs.

####Steps to Reproduce
Unsubscribe a magento customer in dotmailer
Sync
Resubscribe in Magento
Sync importer

#####Actual Result
Importer record remains 'Not Processed' despite being processed

#####Expected Result
Importer record gets set to Processed or Failed so that it's not picked up again

Related work items: #62545

simon-letch committed:
Merged PR 5317: Add release notes for 2.4.2

Add release notes for 2.4.2

simon-letch committed:
Merged PR 5313: 2.4.2 Release

Includes fix to prevent Subscriber_Resubscribed importer messages being processed in a loop.

Related work items: #62545

Commit:
compress and decompress template text between load and save using plugin

Commit:
fix sync for templates when more then one with the same id is mapped

Commit:
make it work with emojy for template subject

Commit:
change the encoding for text body\

Commit:
from name for templates

adeelq committed:
Merged PR 5329: Ignore updates for subscribers which are not active or need confirmation #62576

Ignore updates for subscribers which are not active or need confirmation #62576

Related work items: #62576

adeelq committed:
Merged PR 5307: Move the update of last quote id from abandon cron to campaign cron #62502

Move the update of last quote id from abandon cron to campaign cron #62502

So best way to test is to

-- Create customer with same email across at least two stores/websites.
-- Abandon cart in both stores/websites within few minutes apart. this is to make sure both AC get created in the same cron run.
-- Wait for AC cron to run. Confirm two new AC rows in email_campaign table for same customer but different stores and quotes.
-- Wait for campaign cron to run. Should get two email with correct quote in last_quote_id data field.

Related work items: #62502

Commit:
Merged PR 5306: fix the trans data for orders and clean the additional data included into the

## What is included into the PR to fix :

*  remove productResource from the order collection schema
*  clean the __sleep method as it never hits
*  clean the __wake method as it never hits(removed serilized objects)
* removed any additional data that has no values(null/false/empty) array
* fix for custom attributes when is not mapped/and vice versa - don't need to reset the collection all the time!
* JsonValueIncompatibleWithSchema. A value named 'deliveryAddress'

## Important change :
_INSTEAD OF CASTING THE DATA AND HOPE FOR THE BEST - NOW WE DON'T SEND UNLESS IT HAS EXPECTED VALUE INSIDE THE JSON._

## How to test
Create orders that have delivery/billing address for a customer, also create orders that don't have delivery address.

Result
All orders should be marked as  **imported** inside the importer queue and attached to contact in dotmailer account.

Related work items: #62500, #62569

adeelq committed:
Merged PR 5297: email sending at store level #62529

So found a way which works if email sent DIRECTLY, QUEUE or ADMIN. I did extensive testing on this on all three ways of sending email. I was able to get correct store id every time.

Any additions, changes or feedback is welcome here if anyone thinks i missed something.

Related work items: #62529

Commit:
Merged PR 5415: remove dependency for other modules

Remove dependency for other modules that cause to load the class of disabled module

Related work items: #63053

Commit:
Merged PR 5366: fix the locale for order purchase date

The Problem
Order created date(purchaseDate) is sending the date that is used in default config for the timezone.

The fix
Time is stored in UTC for the orders and there is no need for localization because dotmailer is keeping the dates in the same UTC format.

Also there is no "purchase_date" in M2, we are senging only "purchaseDate" field for orders and this is somehow transformed to match the shema.

Related work items: #58400

adeelq committed:
Merged PR 5455: update readme and up the version number

update readme and up the version number

adeelq committed:
Merged PR 5458: Merge AQ/Master243 to master

Cherry pick commits from testing branch

Related work items: #63053

Commit:
check for template_id as it may be missing

adeelq committed:
Merged PR 5508: bump version number and change it in upgradeData for security changes #62568

bump version number and change it in upgradeData for security changes #62568

Related work items: #62568

Commit:
Merge branch 'testing' into transactionalEmails

# Conflicts:
#	etc/adminhtml/system.xml
#	etc/di.xml

Commit:
Merged PR 5080: Merge transactionalEmails to testing

Commit:
fix conflict after merge the templates branch

Commit:
Merged PR 5546: Merge transactionalEmails to testing

conflict after merging testing branch with templates

Commit:
fix closing element for merge conflict

Commit:
Merged PR 5549: Merge transactionalEmails to testing

cdiacon committed:
Merge pull request #499 from dotmailer/testing

Testing

Commit:
Merged PR 5571: Merge transactionalEmails to testing

Commit:
Merge branch 'transactionalEmails' into testing

Commit:
fix the system xml from the templates branch

Commit:
Merged PR 5572: Merge emailTemplates to testing

Commit:
Merged PR 5573: last working version for system.xml

last working version for system.xml

Commit:
Merged PR 5531: fix the purchase date using locale

fix the purchase date using locale

Commit:
Merged PR 5611: check if enabled before doing api calls

check if enabled before doing api calls

Cron execution to unsubcribe suppressed contacts should check for api enabled before doing api calls.

Currently it checks for the client but the client is always true - so the check fails to stop execution.

Fix should check if api is enabled for each website that is configured to run the sync.

Related work items: #63854

Commit:
Merged PR 5628: fix conlict from the merging with testing branch

fix conlict from the merging with testing branch

Some changes left behind when resolving the conflict with testing branch .

Commit:
Merged PR 5619: Merge calin/default-values/api-calls to develop

Default config path for api requests has the wrong path.

Related work items: #63995

adeelq committed:
Merged PR 5651: Merge AQ/TrackingCode to develop

Related work items: #63852

adeelq committed:
Merged PR 5686: Set min height on fancybox-inner

Set min height on fancybox-inner

adeelq committed:
Merged PR 5683: Fix typos in the Abandoned cart report sections #63306

Fix typos in the Abandoned cart report sections #63306

Related work items: #63306

adeelq committed:
Merged PR 5692: Set the order of the grid items as new one will always be appended to the end

Set the order of the grid items as new one will always be appended to the end #63306

Related work items: #63306

adeelq committed:
Merged PR 5702: Fix for Basket link error #63521

Fix for Basket link error #63521

Related work items: #63521

adeelq committed:
Merged PR 5723: Order sync size is not default to anything #64113

Order sync size is not default to anything #64113

Related work items: #64113

adeelq committed:
Merged PR 5739: Release notes for release

Release notes for release

Commit:
Merged PR 5755: update the readme

update the readme to point to the develop branch for PR

adeelq committed:
Merged PR 5809: Trial account is saving raw password #64470

Trial account is saving raw password #64470

Related work items: #64470

simon-letch committed:
Merged PR 5807: Get campaign from dotmailer with prepared content

This uses the new dotmailer api method which will return a campaign with content prepared for sending via transactional email

Related work items: #64429

kandy commented on Block/Adminhtml/Dashboard/Information.php:
```diff
@@ -143,8 +143,6 @@ public function getDynamicContentPasscode()
-        return ($this->data->getAbandonedCartLimit())?
-            '<span class="message message-warning">' . $this->escapeHtml($this->data->getAbandonedCartLimit()).
-            ' h</span>' : 'No limit';
+        return ($this->data->getAbandonedCartLimit())? $this->data->getAbandonedCartLimit() . ' h': __('No limit');
```
:warning: Please add translate for getAbandonedCartLimit() like __('%1 h',$this->data->getAbandonedCartLimit() )

kandy commented on Controller/Adminhtml/Abandoned.php:
```diff
@@ -0,0 +1,14 @@
+
+namespace Dotdigitalgroup\Email\Controller\Adminhtml;
+
+abstract class Abandoned extends \Magento\Backend\App\Action
```
⚠️ Inheritance is not recommended. Please define ADMIN_RESOURCE constant in every child class

Commit:
Merged PR 5835: Support the templates for store level

Because the templates have the code unique we can't save the same template for different stores.

On unmapping we can't trace what template was removed and we need to delete from email_templates table.

To support the email templates for each store we have to add the store code to the template name ex.

'New Account (dotmailer)__2" <--- 2 will mean that this template was created and mapped for the store with id = 2

Related work items: #64486

kandy commented on Model/Connector/Order.php:
```diff
@@ -638,10 +637,10 @@ public function _getAttributesArray($attributeSetId)
     {
-        if (strlen($value) > 250) {
-            $value = substr($value, 0, 250);
+        if (mb_strlen($value) > 250) {
```
Please use \Magento\Framework\Stdlib\StringUtils here.
Also, use const for 250 magic number

kandy commented on Model/Connector/Product.php:
```diff
@@ -175,8 +175,8 @@ public function setProduct($product)
         //limit short description
-        if (strlen($shortDescription) > 250) {
-            $shortDescription = substr($shortDescription, 0, 250);
+        if (mb_strlen($shortDescription) > 250) {
```
Please use \Magento\Framework\Stdlib\StringUtils here

kandy commented on Model/Connector/Product.php:
```diff
@@ -338,4 +336,18 @@ public function __sleep()
     }
+
+    /**
+     * @param $productAttribute array
```
Add method description

kandy commented on Model/Connector/Product.php:
```diff
@@ -338,4 +336,18 @@ public function __sleep()
+     * @param $productAttribute array
+     * @return string
+     */
+    private function validateProductLabel($productAttribute)
```
This method transform label, not validate. Please rename

simon-letch committed:
Merged PR 5901: Pass required parameters to dotmailer api method

As content is being run through the personalizer, we have to specify an email address to personalise for. It doesn't matter that it doesn't exist. It won't be used.

As content is being compiled and personalized, no need to further convert content. Resized images that were on vedimage will now be on the cdn. Unsub links will be removed. Forward to a friend and cant read will remain, but i think this is ok. We caveat the docs for smtp and api method to say that they won't be replaced or valid. str_replace of "Unsubscribe", "Forward this email" and "View in browser" seems heavy handed as they may want to use these words/phrases in the email.

Related work items: #64426

kandy commented on Model/Cron.php:
```diff
@@ -305,4 +361,17 @@ public function cleaning()
+     * @param $jobCode
+     * @return bool
+     */
+    private function jobHasAlreadyBeenRun($jobCode)
```
Please add description to method and describe why it needs

kandy commented on Model/Importer.php:
```diff
@@ -220,26 +220,34 @@ public function registerQueue(
             $this->helper->debug((string)$e, []);
         }
 
+        if (json_last_error_msg() != "No error") {
```
⚠️ Move this code to place where json_decode is called

kandy commented on Model/Newsletter/Subscriber.php:
```diff
@@ -240,8 +248,10 @@ public function unsubscribe()
-                    $suppressedContact = $apiContact->suppressedContact;
-                    $suppressedEmails[] = $suppressedContact->email;
+                    $suppressedContactEmail = $apiContact->suppressedContact->email;
+                    if (!in_array($suppressedContactEmail, $suppressedEmails)) {
```
Better to use $strict comparison if posible

kandy commented on Model/ResourceModel/Automation/Collection.php:
```diff
@@ -66,4 +66,22 @@ public function getCollectionByType($type, $limit)
+     * @param $automationTypeId
+     * @return bool
+     */
+    public function hasAutomationEnrolmentExists($typeId, $automationTypeId)
```
:exclamation: This method cannot  be called more than one times

adeelq replied:
@kandy Why is this a problem?

kandy replied:
The problem that you should know the implementation of a method to use it. 
Also this violates the 2.12. item of [technical guidelines](http://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html).

adeelq replied:
@kandy We do think you point is right however not for above function. Maybe you can provide us one example to clarify the issue please.

kandy replied:
Example pseudocode:
``` 
$collection =   $this->objectManger->create(\Dotdigitalgroup\Email\Model\ResourceModel\Automation\Collection::class);

if ($collection->hasAutomationEnrolmentExists($typeId, $automationTypeId)) {
     $collection->load(); // <-- will load only one element, but it unexpected result, I did not call any setters  
}
```
To resolve this issue I recommend to extract this type of functions in separate service class

adeelq replied:
Thanks

kandy commented on Model/ResourceModel/Contact.php:
```diff
@@ -819,58 +791,61 @@ private function buildColumnData(
+                LEFT JOIN $eavAttribute ea ON pei.attribute_id = ea.attribute_id
+                LEFT JOIN $eavAttributeOptionValue as eaov on pei.value = eaov.option_id
+                $where
+                AND ea.attribute_code = '$this->brand'
```
⚠️  brand variable  should be properly quoted

kandy commented on Model/ResourceModel/Contact.php:
```diff
@@ -819,58 +791,61 @@ private function buildColumnData(
+     * @param $brand
+     * @return $this
+     */
+    public function setBrandAttribute($brand)
```
⚠️ You should avoid state and temporary coupling in service class

kandy commented on Model/ResourceModel/Contact.php:
```diff
@@ -1034,4 +1009,18 @@ public function getDateLastCronRun($cronJob)
+     *
+     * @param $customerIds
+     */
+    public function setNotImportedByCustomerIds($customerIds)
```
Please update method name to 'update...'

kandy commented on Model/ResourceModel/Cron/Collection.php:
```diff
@@ -19,4 +19,35 @@ public function _construct()
+     * @param $jobCode
+     * @return \Magento\Cron\Model\ResourceModel\Schedule\
+     */
+    public function getRunningJobByCode($jobCode)
```
:exclamation: This method cannot  be called more than one times

adeelq replied:
Same here @kandy Why is this a problem?

kandy replied:
see https://github.com/kandy/dotmailer-magento2-extension/pull/1#discussion_r169621556

kandy commented on Model/ResourceModel/Cron/Collection.php:
```diff
@@ -19,4 +19,35 @@ public function _construct()
+     *
+     * @return bool
+     */
+    public function jobOfSameTypeAndScheduledAtDateAlreadyExecuted($jobCode, $scheduledAt)
```
:exclamation: This method cannot  be called more than one times

adeelq replied:
Same here @kandy Why is this a problem?

kandy replied:
see https://github.com/kandy/dotmailer-magento2-extension/pull/1#discussion_r169621556

kandy commented on Model/ResourceModel/Rules.php:
```diff
@@ -16,6 +22,32 @@ public function _construct()
+     * @param null $connectionName
+     */
+    public function __construct(
+        Json $serializer,
```
Better to preserve the order or argument

adeelq replied:
@kandy Not sure if i understand correctly. Can you please explain more?

simon-letch replied:
The base class constructor takes the args $context, $connectionName in that order. These should be the first two parameters in this class i think so $serializer comes last.

simon-letch replied:
not sure it's possible though with the default argument value for $connectionName?

adeelq replied:
indeed thats why $connectionName is the last param since it has default value

adeelq replied:
@kandy Never mind. I understand what you mean

kandy commented on Model/ResourceModel/Contact.php:
```diff
@@ -819,58 +791,61 @@ private function buildColumnData(
+        $forSubscriber = false
     ) {
+        if (! $this->brand) {
+            return new \Zend_Db_Expr('NULL');
```
Use \Magento\Framework\DB\Sql\Expression here

kandy commented on Model/ResourceModel/Contact.php:
```diff
@@ -819,58 +791,61 @@ private function buildColumnData(
-            );
-        }
-        return $mostData;
+        return $mostData = new \Zend_Db_Expr(
```
Use \Magento\Framework\DB\Sql\Expression here

Commit:
Merged PR 5936: fix emailcapture to make infinite loop calls

Fix emailcapture to make infinite loop calls.

On checkout page the emailcapture is called multiple/infinite times when the password manager is enabled and the user saved the creds to be inserted.

This should fix the multiple calls and checking the previous post do doesn't duplicate the post send if the email has not changed.

Related work items: #65062

Commit:
Merged PR 5982: same template code remove before saving

same template code remove before saving

remove and save again the template as magento is checking for the template with the sam email code.

Related work items: #65183

Commit:
Merged PR 6002: Fix for the default and website using the same template code

fix for the default and website using the same template code

When mapping the templates using the default and website level the store id may have the same ID which also defaults to 0.

Now we should use the default, website or store code when it's present - template name should end with the code depending on what level was mapped.

Related work items: #65245

simon-letch committed:
Merged PR 6043: Spelling correction in 'Create Data Field section of admin site

Default Value comment has 'alwasy' rather than 'always' in its string.

BUNDLE-565

Related work items: #65297

Commit:
Merged PR 6016: Encode/decode compress/decompress only dotmailer templates

encode/decode compress/decompress only dotmailer templates

Compression and decompression should be done only on matching dotmailer templates

Added a regex for template code to match the dotmailer.

Related work items: #65261

simon-letch committed:
Merged PR 6039: Don't unbind when email is empty, invalid or not changed. If we do, the blur

Don't unbind when email is empty, invalid or not changed. If we do, the blur event is not listened to ever again.

This rolls back to the simpler version we had before. It now saves the last received email, if it's not changed since last blur, then return false as it's likely we're in an infinite loop caused by browser auto fills.

Related work items: #65062

Commit:
Merged PR 6055: Merge calin/email-templates-iheritance to develop

When syncing template the default store/website is still using the ID for the code name and the wrong template will be synced.

This should replace the store ID with 'admin' code name.
Ignore Inherited options for the template if was not selected. By default we removed the ignore for inherited items so it will not be saved against the store but the parent option. Now we put back the check for inhereted items.

Related work items: #65352

Commit:
Merged PR 6109: Merge calin/email-templates-cron-sync to develop

This should contain and fix the template naming for each store level. Template sync will sync once for each website the campaigns with the same "campaign id".

Fixed the regular expresion for new templates name change and should ignore to include into the sync core templates.

Related work items: #65347

adeelq committed:
Merged PR 6013: update read me

update read me

adeelq committed:
Merge branch 'develop' into bundling-2.2.4

# Conflicts:
#	README.md
#	composer.json
#	etc/module.xml

kandy commented on Block/Adminhtml/Config/Developer/EmailTemplatesSync.php:
```diff
@@ -0,0 +1,44 @@
+     * @return string
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function _getElementHtml(\Magento\Framework\Data\Form\Element\AbstractElement $element)
```
Mark method  as protected

kandy commented on Block/Adminhtml/Config/Developer/Templatesync.php:
```diff
@@ -0,0 +1,44 @@
+     * @return string
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function _getElementHtml(\Magento\Framework\Data\Form\Element\AbstractElement $element)
```
Mark method protected

adeelq committed:
Merged PR 6226: Merge AQ/SecurityReview to develop

Related work items: #65786, #65795, #65802, #65819

adeelq committed:
Merged PR 6203: architect review

Following comments left todo. waiting for response.

This method cannot be called more than one times
Better to preserve the order or argument

Related work items: #65759

Commit:
Merged PR 6294: code style for plugins

This what the code sniffer violation are suggesting to change we have more but most of them are already marked and agreed that WON'T FIX.

Related work items: #65956

simon-letch committed:
Fixes for newsletter subscription options

When 'Show datafields to customer' is set to false, the request does not contain data_fields. This causes an error when a user updates their additional newsletter subscriptions. The change to getContactDataFields checks if it's set, if not returns an empty array. If it has been set, but is an empty array, it returns an empty array too. This saves an additional call to dotmailer.

As per github 502, we weren't correctly updating preferences. Now we iterate over the configured additional address books and maintain those mappings. We don't touch other address book subscriptions.

https://github.com/dotmailer/dotmailer-magento2-extension/issues/501
https://github.com/dotmailer/dotmailer-magento2-extension/issues/502

Related work items: #65828

adeelq committed:
Merged PR 6272: Merge AQ/MFTF to develop

Related work items: #65950, #65956, #65980

adeelq committed:
Merged PR 6286: Merge AQ/SecurityReview to develop

Related work items: #65798, #65921, #66073

adeelq committed:
Merged PR 6365: Increase secret key length #66097

Increase secret key length #66097

Related work items: #66097

simon-letch committed:
Merged PR 6353: Regionify tracking script. Get it from config or default to t.trackedlink.net

Regionify tracking script.

Get it from config or default to t.trackedlink.net when no region can be determined

Related work items: #56353

adeelq committed:
Merged PR 6369: Merge AQ/ArchitectReview to develop

Related work items: #65759

Commit:
Merged PR 6368: check and create abandoned carts table

Abandoned cart table is not create in the upgrade script when upgrading from 2.3.8

When upgrading from a bundling version there are no abandoned cart feature and the missing table error for abandoned cart will trigger.

This should fix the error when abandoned cart table was missing inside the recurring script. Recurring file will be executed every time when the setup:upgrade command runs and will check that abandoned cart table should be created on version number 2.3.8 and up.

Assuming that he next versions of the bundle will have the same issue we have to keep on checking everytime the setup upgrade runs.

Related work items: #66277

simon-letch committed:
Stop sending AC emails if cart has been emptied

AC1 was being sent out, customer went back to the cart, emptied the contents. They still received AC2.

Previously we were doing quote item count checks, but since the refactor we do not.

The quote collection was not picking up any quotes with a item count of 0, so even though the quote was updated, the AC1 logic did not pick it up. The AC2/3 logic queries our AC table and takes no notice of items count. We can catch empty baskets at AC1 by removing the items_count and active predicates. If we detect an empty quote or inactive one, then i've changed the logic to not send any email and delete the AC row in our table if it exists.

I've also fixed the typo in process.

I've put the logic to check item count and active status in AC1, 2 and 3 logic. That way, should another rogue extension update/inactivate carts without updating the updated_at date we'll still not send out empty emails.

Related work items: #66202

Commit:
Merge branch 'develop' into bundling-2.2.4
